stdlib: add Simulator API to schedule hypercall exits

erin-le opened the pull request:
This commit adds functions to Simulator to allow for m5.scheduleTickExitAbsolute and m5.scheduleTickExitFromCurrent to be used in a config script without importing m5. By extension, this allows for hypercall 6 to be trigged and hypercall scheduled exit handlers to be used without having to import m5 into config scripts.

erin-le committed:
stdlib: add Simulator API to schedule hypercall exits

This commit adds functions to Simulator to allow for
m5.scheduleTickExitAbsolute and m5.scheduleTickExitFromCurrent
to be used in a config script without using m5. By extension, this
allows for hypercall 6 to be trigged and hypercall scheduled exit
handlers to be used without having to import m5 into config
scripts.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -282,12 +283,71 @@ def set_max_ticks(self, max_tick: int) -> None:
         return self._max_ticks
 
+    def set_hypercall_absolute_max_ticks(
+        self, max_tick: int, exit_str: str
```
How about 
```suggestion
        self, max_tick: int, exit_str: str="Max ticks reached"
```

And you can drop the "This must be provided to exit with hypercall 6 exit."

In the long run, I'd like to deprecate the old way of doing it, and I think having a default string will make it easier.

Maybe also `warn` if the string is empty.

erin-le replied:
I made the suggested changes and tested. If the user passes "" for `exit_str`, the function will still trigger hypercall 6. However, if they use "Tick exit reached", which is the default string in m5.scheduleTickExitAbsolute and m5.scheduleTickExitCurrent, they will trigger `ExitEvent.SCHEDULED_TICK`. As such, I've added a warning for if users accidentally use "Tick exit reached".

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -282,12 +283,71 @@ def set_max_ticks(self, max_tick: int) -> None:
+        self._hypercall_max_ticks = max_tick
+
+    def set_hypercall_relative_max_ticks(
+        self, ticks_from_current: int, exit_str: str
```
Ditto above

erin-le committed:
stdlib: add default string to hypercall 6 API in Simulator

This commit adds a default string to the functions that allow
users to trigger hypercall 6 via Simulator. This removes the
requirement for users to pass in an exit string when using
these functions.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -282,12 +283,79 @@ def set_max_ticks(self, max_tick: int) -> None:
+        m5.scheduleTickExitAbsolute(max_tick, exit_str)
+        self._hypercall_max_ticks = max_tick
+
+    def set_hypercall_relative_max_ticks(
```
Sorry one other thing... why don't you just call `set_hypercall_absolute_max_ticks(curTick() + ticks_from_current)` and remove the duplicated warnings.

erin-le committed:
stdlib: modify hypercall relative max ticks API in Simulator

This commit modifies Simulator.set_hypercall_relative_max_ticks
to use Simulator.set_hypercall_absolute_max_ticks instead of
m5.scheduleTickExitFromCurrent. This change removes duplicate warning
messages.

powerjg commented on src/python/gem5/simulate/simulator.py:
```diff
@@ -282,12 +283,65 @@ def set_max_ticks(self, max_tick: int) -> None:
+        """
+        max_tick = self.get_current_tick() + ticks_from_current
+        self.set_hypercall_absolute_max_ticks(max_tick, exit_str)
+        self._hypercall_max_ticks = max_tick
```
This line seems redundant with line 326. Let me know if I'm missing something.

erin-le committed:
stdlib: remove redundant line

This commit removes a redundant line from
Simulator.set_hypercall_relative_max_ticks, as it would already
be set by the call to set_hypercall_absolute_max_ticks.
